datadog_checks_base/datadog_checks/base/utils/concurrency/limiter.py

```diff
@@ -0,0 +1,76 @@
+# (C) Datadog, Inc. 2022-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import threading
+
+
+class ConditionLimiter(object):
+    """
+    This class is used to limit the number of concurrent satisfied conditions. Every subclass must implement
+    a `condition` method. For example, if you wanted to only run one check instance at a time based on the
+    command line arguments used to start an active process, you could do:
+
+    ```python
+    import re
+    import shlex
+
+    import psutil
+
+    from datadog_checks.base import AgentCheck
+    from datadog_checks.base.utils.concurrency.limiter import ConditionLimiter
+
+    class Limiter(ConditionLimiter):
+        def condition(self, pattern, logger):
+            logger.info('Searching for a process that matches: %s', pattern)
+            for process in psutil.process_iter(['cmdline']):
+                command = shlex.join(process.info['cmdline'])
+                if re.search(pattern, command):
+                    logger.info('Process found: %s', command)
+                    return True
+            else:
+                logger.info('Process not found, skipping check run')
+                return False
+
+    class Check(AgentCheck):
+        limiter = Limiter()
+
+        def check(self, _):
+            if not self.limiter.check_condition(self.check_id, self.config.pattern, self.log):
+                return
+
+            try:
+                ...
+            except Exception:
+                self.limiter.remove(self.check_id)
+                raise
+    ```
+    """
+
+    def __init__(self, limit=1):
+        self.__limit = max(limit, 1)
+        self.__cached_identifiers = set()
+        self.__lock = threading.Lock()
+
+    def limit_reached(self):
+        return len(self.__cached_identifiers) >= self.__limit
+
+    def remove(self, identifier):
+        return self.__cached_identifiers.remove(identifier)
+
+    def check_condition(self, identifier, *args, **kwargs):
+        if self.limit_reached():
+            return identifier in self.__cached_identifiers
+
+        with self.__lock:
+            # Check if the limit has been reached while locked
+            if self.limit_reached():
+                return identifier in self.__cached_identifiers
+
+            condition_met = self.condition(*args, **kwargs)
```

Comment:
This would run the condition every time if limit is greater than the number of instances. I assume it is expected to be expensive, otherwise checks would just run condition every time. Would it make sense to skip subsequent runs once it succeeds for a given identifier?

Reply:
Fixed!